src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -170,7 +171,13 @@ String getClassNameInternal() {
         cacheKey = new CityHash128Key(sql);
 
         // Parse or fetch SQL metadata from cache.
-        ParsedSQLCacheItem parsedSql = getOrCreateCachedParsedSQL(cacheKey, sql);
+        ParsedSQLCacheItem parsedSql = getCachedParsedSQL(cacheKey, sql);
+        if (null != parsedSql) {
+            isExecutedAtLeastOnce = true;
+        }
+        else {
+            parsedSql = parseAndCacheSQL(cacheKey, sql);
+        }
```

Comment:
Just to clarify, as far as I see it this would be the only difference between the two implementations, i.e. at the end of the day you will get handle re-use in both. For the "single call on connection" case the connection level is better, for the other cases I would say difference will be negligible:
```
1 call
sp_executesql | sp_prepexec
              | sp_unprepare

2 calls
sp_executesql | sp_prepexec
sp_prepexec   | sp_execute 
sp_unprepare  | sp_unprepare

3+ calls
sp_executesql | sp_prepexec
sp_prepexec   | sp_execute 
sp_execute    | sp_execute
...           | ...
sp_unprepare  | sp_unprepare
```